### PR TITLE
Implement option: view_df

### PR DESCRIPTION
**Pull request by jalvesaq**

### CODE REVIEW

**Comment by jalvesaq** [comment]
Following the discussion on #262, in this pull request the options to view a `data.frame` are in a new table: `view_df` (please, see the documentation). I tried some combinations of values, but it will be better if other people review the changes. Below are some examples:

**Show only the first five lines in a vertical split window using semicolon as field separator:**

```lua
view_df = {
    csv_sep = ";",
    n_lines = 5,
    how = "vsplit",
},
```

**Use `fwrite()` to quickly write the CSV file and display all lines in VisiData regardless of the `data.frame` size:**

```lua
view_df = {
    n_lines = 0,
    save_fun = "data.table::fwrite",
    open_app = "terminal:vd",
},
```

In the `~/.Rprofile`:

```r
options(datatable.fwrite.sep = "\t")
```

**Use `tibble()` to display the `data.frame` on R Console:**

```lua
view_df = {
    open_fun = "tibble::tibble",
},
```

**Comment by PMassicotte** [comment]
From the description, I really like it. I am teaching today, but will be able to look at the code tomorrow. Thank!

**Comment by PMassicotte** [comment]
I am trying the different options you are proposing.

```lua
view_df = {
    csv_sep = ";",
    n_lines = 5,
    how = "vsplit",
},
```

gives me an empty split:

image

```lua
                view_df = {
                    n_lines = 0,
                    save_fun = "data.table::fwrite",
                    open_app = "terminal:vd",
                },
```

Gives me this error:

```r
[ins] r$> Error in nvimcom:::nvim_viewobj("mtcars", nrows = 0, save_fun = data.table::fwrite) : 
  unused argument (save_fun = data.table::fwrite)
```

```lua
view_df = {
    open_fun = "tibble::tibble",
},
```

works fine.

File: doc/R.nvim.txt [hunk 1]
```diff
@@ -1843,64 +1843,96 @@ See https://github.com/R-nvim/R.nvim/wiki/Remote-access
...
+      open_app = "",  -- How to open the CSV
+      how = "tabnew", -- How to display the data if doing it within Neovim
+      csv_sep = "",   -- Field separator to be used when saving the CSV.
+      n_lines = -1,   -- Number of lines to save in the CSV (0 for all lines).
+      save_fun = "",  -- R function to save the data.frame in a CSV file
+      open_fun = "",  -- R function to open the data.frame directly
+                      -- (no conversion to CSV needed)
+   },
+<
+If `open_app` begins with "terminal:", R.nvim will open a new tab with the
```
Comment (PMassicotte) [inline doc/R.nvim.txt:1864]: We are talking about TSV and then all the options are about CSV. Maybe make it clearer?

File: doc/R.nvim.txt [hunk 2]
```diff
@@ -1843,64 +1843,96 @@ See https://github.com/R-nvim/R.nvim/wiki/Remote-access
...
 in your config. Valid values are `"\t"`, `";"` and `","`. Example:
 >lua
-   csv_delim = ","
+   csv_sep = ","
+<
+By default, the `tsv` file is got by capturing the output of `write.table()`.
+This has the advantage of depending only on base R. The data is streamed
+through nvimcom's TCP connection to ensure that it will work if R is running
+remotely. If this is too slow for the size of your data sets, or if you want
+to display more lines of data, you will want to change other `view_df`
```
Comment (PMassicotte) [inline doc/R.nvim.txt:1910]:
```suggestion
to display more lines of data, you will want to change other `view_df` by using another R function like `save_fun = data.table::fwrite`.
```

File: lua/r/config.lua [hunk 3]
```diff
@@ -504,6 +503,14 @@ local config = {
     tmpdir              = "",
     user_login          = "",
     user_maps_only      = false,
+    view_df = {
+        open_app = "",  -- How to open the CSV in Neovim or an external application.
+        how = "tabnew", -- How to display the data within Neovim if not using an external application.
+        csv_sep = "",   -- Field separator to be used when saving the CSV.
```
Comment (PMassicotte) [inline lua/r/config.lua:509]:
```suggestion
        csv_sep = "",   -- Field separator to be used when saving the TSV.
```
Should standardize TSV/CSV everywhere.

File: doc/R.nvim.txt [hunk 4]
```diff
@@ -1843,64 +1843,96 @@ See https://github.com/R-nvim/R.nvim/wiki/Remote-access
...
+`()` are added if necessary. In this case, no `csv` file is saved (see also
+|R.nvim-key-bindings|).
 
-If `csv_app` is not defined, there will be addional maps to open the file in a
-split window, in a vertical split window, or to see only the head of the csv
-in a split window.
+The data will be displayed in a new tab. Change the value of `how` to "split"
+or "vsplit" to display the data in a split window.
 
 If the default field delimiter causes problems in your case, you can change it
```
Comment (PMassicotte) [inline doc/R.nvim.txt:1901]:
```suggestion
If the default field delimiter (`\t`) causes problems in your case, you can change it
```

File: doc/R.nvim.txt [hunk 5]
```diff
@@ -1843,64 +1843,96 @@ See https://github.com/R-nvim/R.nvim/wiki/Remote-access
...
-There is also the option of configuring R.nvim to run an R command to
+Set the value of `n_lines` to `0` to display all lines or to any positive
+integer number to set a fixed maximum.
+
+If are not running R remotely and want the CSV file written more quickly, you
+can define the function to save the CSV. In this case, the data will be saved
+in a file instead of streamed. The function must receive the `data.frame` as
+the first argument and the file name as the second argument. For example, if
+you have the `data.table` library installed and put in your `~/.Rprofile`
+`options(datatable.fwrite.sep = "\t")`, you could use the function `fwrite`:
```
Comment (PMassicotte) [inline doc/R.nvim.txt:1925]: Is this needed since there is the  `csv_sep`  option?

**Comment by PMassicotte** [comment]
> ```lua
>                 view_df = {
>                     n_lines = 0,
>                     save_fun = "data.table::fwrite",
>                     open_app = "terminal:vd",
>                 },
> ```
> 
> Gives me this error:
> 
> ```r
> [ins] r$> Error in nvimcom:::nvim_viewobj("mtcars", nrows = 0, save_fun = data.table::fwrite) : 
>   unused argument (save_fun = data.table::fwrite)
> ```

I think it works now, for some reason, nvimcom was not rebuilt...

**Comment by PMassicotte** [comment]
This is pretty fast with `data.table`, :+1:

**Comment by jalvesaq** [comment]
> I think it works now, for some reason, nvimcom was not rebuilt...

`nvimcom` is rebuilt only when its version number changes, but I usually only change the version number once during the development of a PR. So, if there is any change in `nvimcom`'s code, it's necessary to do the following:

```
echo 'remove.packages("nvimcom")' | R
```

Maybe I should change this procedure. Otherwise, `nvimcom` will be rebuilt only when you switch branches.

**Comment by jalvesaq** [comment]
We have to decide about saving the file as CSV or TSV and naming it as CSV or TSV, and about the names of the options. I'm using "CSV" as a generic name for any text file with data separated by a delimiter (usually, commas, semicolons, or tabs). Vim plugins and tree-sitter parsers have "csv" in their names, but they also support TSV. "CSV" is the most common name, but in Brazil, the default is to use semicolons because the decimal separator is a comma. "TSV" corresponds to tab-delimited and is the most convenient to open because it doesn't require quotes around text.

**Comment by jalvesaq** [comment]
> Is this needed since there is the csv_sep option?

Without this option, `fwrite()` will save a comma-delimited file (which seems to be a bit slower than saving a TSV file), and `vd` will not recognize the file format although the extension is `.csv`. VisiData seems to be geared towards TSV, not CSV.

File: doc/R.nvim.txt [hunk 6]
```diff
@@ -1843,64 +1843,96 @@ See https://github.com/R-nvim/R.nvim/wiki/Remote-access
```
[the remark this answers is not in the document]
Comment (jalvesaq) [reply doc/R.nvim.txt:1843]: No. This is old. But the default implementation of how to view a `data.frame` is limited to make it work even when R is running remotely. That's why the text of the TSV is streamed through the TCP connection. When Neovim and R are on the same machine, it's easier to make R save the TSV and Neovim open it.

**Comment by PMassicotte** [comment]
> We have to decide about saving the file as CSV or TSV and naming it as CSV or TSV, and about the names of the options. I'm using "CSV" as a generic name for any text file with data separated by a delimiter (usually, commas, semicolons, or tabs). Vim plugins and tree-sitter parsers have "csv" in their names, but they also support TSV. "CSV" is the most common name, but in Brazil, the default is to use semicolons because the decimal separator is a comma. "TSV" corresponds to tab-delimited and is the most convenient to open because it doesn't require quotes around text.

Sounds good to me. Been tested it for 1 hour and all fine from what I can see.

**Comment by kraxli** [comment]
thanks again @jalvesaq  for all the work! I have played with the `branch = 'view_df',` and observed the following: 

Using: 

```lua
save_fun = "data.table::fwrite",
```
saves a `tsv` file but comma separated. Hence, `vd` is not able do detect the format, and the file is shown with one column only:

image

Additionally, I have tried to define my own `arrow` based function:

```lua
save_fun = 'function(df) {arrow::write_dataset(df, path = ".", basename_template = paste0(deparse(substitute(df)), ".", as.character(format)), format = "parquet")}',
```

which results in
```R
> Error in save_fun(o, paste0(oname, ".csv")) : 
  unused argument (paste0(oname, ".csv"))
```

**Comment by jalvesaq** [comment]
Please, see the example in the documentation and in my first comment here. You have to put in your `~/.Rprofile`:

```r
options(datatable.fwrite.sep = "\t")
```

**Comment by jalvesaq** [comment]
`save_fun` is the name of an R function, not a complete expression to be evaluated. It would be interesting if it could be a complete expression, and a pull request for this could be merged.

**Comment by jalvesaq** [comment]
Note: I tested and it works with complete expressions. I will give some examples soon.

**Comment by jalvesaq** [comment]
I fixed some issues and this should work now:

**Use `arrow` to save a parquet file, and VisiData to see it:**

```lua
view_df = {
    save_fun = "function(obj, obj_name) { arrow::write_dataset(obj, path = '.', basename_template = paste0(obj_name, '{i}.parquet'), format = 'parquet') ; return(paste0(obj_name, '0.parquet')) }",
    open_app = "terminal:vd"
},
```

**Comment by jalvesaq** [comment]
I'm starting to think that this feature is becoming too complicated. My new suggestion:

1. Always output a TSV file.
2. Use `data.table::fwrite` to output the TSV file if `data.table` is installed; otherwise, use `utils::write.table`.
3. Remove the options `csv_sep` and `save_fun`.

Is there any advantage of using the parquet format?

**Comment by jalvesaq** [comment]
Although complicated, it seems to work. I tried these examples:

**Show only the first five lines in a vertical split window using semicolon as field separator:**

```lua
view_df = {
    csv_sep = ";",
    n_lines = 5,
    how = "vsplit",
},
```

**Use tibble() to display the data.frame on R Console:**

```lua
view_df = {
    open_fun = "tibble::tibble",
},
```

**Use a Lua function to see the first five lines as a notification:**


```lua
view_df = {
    n_lines = 5,
    open_app = function(file_name)
        local lines = vim.fn.readfile(file_name)
        vim.notify(table.concat(lines, "\n"))
    end
},
```


**Use `fwrite()` to quickly export the `data.frame` to a CSV file and open it in VisiData:**

```lua
view_df = {
    n_lines = 0,
    save_fun = "data.table::fwrite",
    open_app = "terminal:vd",
},
```

In the `~/.Rprofile`:

```r
options(datatable.fwrite.sep = "\t")
```

**Use `fwrite()` without set any option in the `~/.Rprofile`:**

```lua
view_df = {
    n_lines = 0,
    save_fun = "function(obj, obj_name) {f <- paste0(obj_name, '.tsv'); data.table::fwrite(obj, f, sep = '\t') ; f}",
    open_app = "terminal:vd",
},
```

**Use `arrow` to save a parquet file, and VisiData to open it:**

```lua
view_df = {
    n_lines = 0,
    save_fun = "function(obj, obj_name) {f <- paste0(obj_name, '.parquet'); arrow::write_parquet(obj, f) ; f}",
    open_app = "terminal:vd"
},
```
**Use `nanoparquet` to save a parquet file, and VisiData to open it:**

```lua
view_df = {
    n_lines = 0,
    save_fun = "function(obj, obj_name) {f <- paste0(obj_name, '.parquet'); nanoparquet::write_parquet(obj, f) ; f}",
    open_app = "terminal:vd"
},
```

**Comment by kraxli** [comment]
I have e.g. tried

```lua
view_df = {
    n_lines = 0,
    save_fun = "function(obj, obj_name) {f <- paste0(obj_name, '.parquet'); arrow::write_parquet(obj, f) ; f}",
    open_app = "terminal:vd"
},
```
on the `flights` data. The file written is `flights.csv.parquet`. Additionally, `vd` does not (try to) open a file, but opens just the directory (maybe because `flights.parquet` does not exist).

**Comment by jalvesaq** [comment]
Maybe you need to remove `nvimcom` to force its update:

```
echo 'remove.packages("nvimcom")' | R
```

I will change nvimcom's version number anyway to ensure its update...

**Comment by PMassicotte** [comment]
All work fine on my side.

**Comment by PMassicotte** [comment]
This is the perfect solution for me:

```lua
view_df = {
                    n_lines = 0,
                    save_fun = "function(obj, obj_name) {f <- paste0(obj_name, '.parquet'); nanoparquet::write_parquet(obj, f) ; f}",
                    open_app = "terminal:vd",
                },
```

**Comment by kraxli** [comment]
> Maybe you need to remove `nvimcom` to force its update:
> 
> ```
> echo 'remove.packages("nvimcom")' | R
> ```
> 
> I will change nvimcom's version number anyway to ensure its update...

right - thanks @jalvesaq! Sorry somewhere in the chat trail you have made this point already.

```lua
        view_df = {
            n_lines = 0,
            save_fun = "function(obj, obj_name) {f <- paste0(obj_name, '.parquet'); arrow::write_parquet(obj, f) ; f}",
            -- open_app = "terminal:vd",
            open_app = ':TermExec cmd="vd %s" direction=float name=visidataTerm',
        },
```
works like a charm :-) and is super fast also for bigger data.frames

**Comment by jalvesaq** [comment]
Thank you! I will merge it now!

**Pull request merged**